packages/live-share-inking/src/canvas/DryWetCanvas.ts

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the Microsoft Live Share SDK License.
+ */
+
+import { InkingCanvas } from "./InkingCanvas";
+import { getPressureAdjustedSize, IPointerPoint, DefaultPenBrush, IBrush, toCssColor } from "../core";
+import { computeQuadBetweenTwoCircles, computeQuadBetweenTwoRectangles, IQuadPathItem } from "../core/Internals";
+
+/**
+ * Represents the base class from wet and dry canvases, implementing the common rendering logic.
+ */
+export abstract class DryWetCanvas extends InkingCanvas {
+    private _innerLayer?: CanvasRenderingContext2D;
+    private _pendingPointsStartIndex = 0;
+    private _points: IPointerPoint[] = [];
+
+    private computeQuadPath(tipSize: number): IQuadPathItem[] {
+        const result: IQuadPathItem[] = [];
+        const tipHalfSize = tipSize / 2;
+
+        if (this._pendingPointsStartIndex < this._points.length) {
+            let previousPoint: IPointerPoint | undefined = undefined;
+            let previousPointPressureAdjustedTip = 0;
+
+            if (this._pendingPointsStartIndex > 0) {
+                previousPoint = this._points[this._pendingPointsStartIndex - 1];
+                previousPointPressureAdjustedTip = getPressureAdjustedSize(tipHalfSize, previousPoint.pressure);
+            }
+
+            for (let i = this._pendingPointsStartIndex; i < this._points.length; i++) {
+                const p = this._points[i];
+
+                let pressureAdjustedTip = getPressureAdjustedSize(tipHalfSize, p.pressure);
+
+                const pathItem: IQuadPathItem = {
+                    endPoint: p,
+                    tipSize: pressureAdjustedTip
+                };
+
+                if (previousPoint !== undefined) {
+                    pathItem.quad = this.brush.tip === "ellipse"
```

Comment:
Should this use BrushTipShape?

Reply:
It "does" :) BrushTipShape isn't an enum, it's a type that defines valid values. This is all type checked by TypeScript.